server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -64,11 +64,14 @@ public static final class VersionLookup {
 
         // Modifies the map of this instance by merging with the given VersionLookup
         public void merge(VersionLookup versionLookup) {
-            map.putAll(versionLookup.map);
+            for (var entry : versionLookup.map.entrySet()) {
+                put(entry.getKey(), entry.getValue());
+            }
```

Comment:
This works functionally. But I am not sure whether `putAll`'s presizing has any important performance difference compared to calling `put` in a loop.

Reply:
This is not map#put but VersionValue#put. The change is on purpose to trigger the memory accounting logic. I agree that performance-wise, it has a disadvantage. For context, this will get triggered once we have a second (and further) local-refresh w/o a flush, since they would both go to the same generation. However, this only impacts stateless ES. we COULD use putAll and do the bookkeeping separately. I think it might be worth optimizing.

Reply:
one complication might be that because we call merge in the archive outside the synchronization block, optimizing this needs a bit more care.

Reply:
Looking at this again, i think the sync block in archive#afterRefresh is not needed for a safe merge. Those afterRefresh calls are triggered by Lucene's ReferenceManager which seems to already use a refresh lock. @henningandersen is the put vs putAll something you think needs optimizing in this PR? Theoretically it would make a difference but as I mentioned I think it would highly depend on the number of local refreshes between flushes and how many new entries it would see, which is a bit hard to generalize.

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/102752/commits/a1396164b0076c215ad22cfdd19d8b7a12a82065.

Reply:
> Those afterRefresh calls are triggered by Lucene's ReferenceManager which seems to already use a refresh lock.

I think so. Otherwise we should already be seeing problems.

Reply:
I think we can defer the optimization. We already populate the other map with the things we merge in. The additional resizes here are likely less harmful than the original ones, since CHM uses a 2x resizing strategy. So I'd be fine to leave this and only address it if it turns out important (which I doubt). We only had putAll to begin with because it was convenient, not because it was important I think.

Reply:
No, I only did not read it in that order, I think it is fine either way.